docs/graphql/manual/api-reference/graphql-api/query.rst

```diff
@@ -120,6 +120,7 @@ E.g.
       address(path: "city") # scalar JSON field -> property; '$.' prefix is optional
       contacts(path: "[0]") # scalar JSON field -> array_item
       contacts(path: "[0].phone") # scalar JSON field -> array_item_property
+      contacts(path: "[Hello world!]") # scalar JSON field -> property; used for special characters key
```

Comment:
This syntax doesn’t seem quite right to me. Both the `a` and `[0]` syntax reflect the syntax used in JavaScript for accessing properties:

```js
> let obj = { a: 1, b: 2 }
> obj.a
1
> let arr = [1, 2]
> obj[0]
1
```

But `[Hello, world!]` isn’t valid JavaScript syntax. The correct syntax should be `["Hello, world!"]`:

```js
> let obj = { ["Hello, world!"]: 1 }
> obj["Hello, world!"]
1
```

Leaving out the quotes actually creates an ambiguity in the grammar, since it makes it impossible to access a numeric object field such as this one:

```json
{ "0": "some value" }
```

So we should use an extra layer of quoting here, just like JavaScript does.

Reply:
@lexi-lambda I agree with your opinion. However, we can't use quote `"`. It will be harder to read and parse, because GraphQL argument use it already `(path: "abc")`. It will break GQL specs.

We can use single quote `'` for this case. JSON path also use this style.
https://github.com/json-path/JsonPath
`$['store']['book'][0]['author']`

Reply:
[GraphQL string literals support escaping.](http://spec.graphql.org/June2018/#StringValue) Moreover, the parsing logic should be exactly the same amount of effort, since the `graphql-parser` library already does the work of parsing GraphQL string syntax, so escapes have already been handled at that phase. Once we get the string in `graphql-engine`, the backslashes will already be gone.

That said, your point about it being harder to read and write for humans is valid. Writing `"$['store']"` looks nicer than writing `"$[\"store\"]`. I picked `"` originally because JSON supports `"` but not `'`, so it seemed more appropriate, but I don’t see any reason not to support `'` as well. Let’s just support both—it’s trivial to do if we’ve already supported one!